src/live/src/user-to-user/stream.js

```diff
@@ -0,0 +1,245 @@
+   * @returns {Promise}
+   */
+  _subscribe(userId, receiver) {
+    return this._requestSubscribeAccess(userId)
+      .then(() => {
+        const channelName = this.subscribeChannels[userId] || this._buildChannelName(userId);
+        return this.liveService.subscribeToChannel(channelName, receiver);
+      });
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @returns {Promise<void>}
+   */
+  _unsubscribe(userId) {
+    return this._unsubscribeFromSubstream(userId)
+      .then(() => {
+        const channelName = this.subscribeChannels[userId] || this._buildChannelName(userId);
+        this.liveService.unsubscribeFromChannel(channelName);
+      });
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @param {Object} message
+   * @returns {Promise}
+   */
+  _publish(userId, message) {
+    if (this.publishChannels[userId]) {
+      return this._retriablePublish(userId, message);
+    }
+    return this._publishWithAccessRequest(userId, message);
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @param {Object} message
+   * @returns {Promise}
+   */
+  _publishWithAccessRequest(userId, message) {
+    return this._requestPublishAccess(userId)
+      .then(() => {
+        const channelName = this.publishChannels[userId] || this._buildChannelName(userId);
+        return this.liveService.publishToChannel(channelName, message);
+      });
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @param {Object} message
+   * @returns {Promise}
+   */
+  _retriablePublish(userId, message) {
+    const channelName = this.publishChannels[userId] || this._buildChannelName(userId);
+    return this.liveService.publishToChannel(channelName, message)
+      .catch((err) => {
+        let promise = Promise.reject(err);
+        if (err.statusCode === StatusCode.Forbidden) {
+          promise = this._publishWithAccessRequest(userId, message);
+        }
+        return promise;
+      });
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @returns {string} The name of the channel for that user
+   */
+  _buildChannelName(userId) {
+    return `${this.client.appKey}.s-${this.name}.u-${userId}`;
+  }
+
+  /**
+   * @private
+   * @param {string} substreamOwnerId
+   * @returns {Promise}
+   */
+  _requestPublishAccess(substreamOwnerId) {
+    return this._makeStreamRequest(`${substreamOwnerId}/publish`, RequestMethod.POST)
+      .then((resp) => {
+        if (resp.substreamChannelName) {
+          this.publishChannels[substreamOwnerId] = resp.substreamChannelName;
+        }
+        return resp;
+      });
+  }
+
+  /**
+   * @private
+   * @param {string} userId
+   * @returns {Promise}
+   */
+  _requestSubscribeAccess(userId) {
+    const requestBody = { deviceId: this.client.deviceId };
+
+    return this._makeStreamRequest(`${userId}/subscribe`, RequestMethod.POST, requestBody)
+      .then((response) => {
+        this.subscribeChannels[userId] = response.substreamChannelName;
```

Comment:
Answer is almost the same here - gets overwritten every time `_requestSubscribeAccess()` is called, which is every time `_subscribe()` is called, which is whenever any of the subscription methods are called - `listen()`, `follow()`.

Difference from `publishChannels` is that `_requestSubscribeAccess()` is called every time, since subscribing doesn't impact performance as much and one additional HTTP call isn't that big of a deal, especially since PubNub timetokens let you receive last 100 (or 200 was it) messages, even if you weren't subscribed to that particular channel, just the channel group.

Perhaps we should discuss this, since we are using a single channel group and it might happen that you subscribe and immediately get 100 messages. This might not be expected. Some manual work with the timetokens might be required, if we wish to avoid this.